From: Juergen Gross <jgross@suse.com>
Subject: Xen: Use machine addresses in /sys/kernel/vmcoreinfo when PV
References: bsc#1014136
Patch-mainline: 4.12
Git-commit: 29985b09613ba106a1ed0496988636d288600515

The values from the sysfs kernel/vmcoreinfo file are used to construct
ELF headers which are used by the kexec'ed kernel, i.e. after the Xen
hypervisor crashes and launches a new OS.

The kexec'ed kernel runs on bare metal, so it requires machine addresses,
and not physical addresses of a (no longer existing) Dom0 kernel.

Signed-off-by: Juergen Gross <jgross@suse.com>
Signed-off-by: Petr Tesarik <ptesarik@suse.com>
---
 arch/x86/xen/mmu.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

--- a/arch/x86/xen/mmu.c
+++ b/arch/x86/xen/mmu.c
@@ -48,6 +48,9 @@
 #include <linux/memblock.h>
 #include <linux/seq_file.h>
 #include <linux/crash_dump.h>
+#ifdef CONFIG_KEXEC_CORE
+#include <linux/kexec.h>
+#endif
 
 #include <trace/events/xen.h>
 
@@ -2932,3 +2935,13 @@ int xen_unmap_domain_gfn_range(struct vm
 #endif
 }
 EXPORT_SYMBOL_GPL(xen_unmap_domain_gfn_range);
+
+#ifdef CONFIG_KEXEC_CORE
+unsigned long paddr_vmcoreinfo_note(void)
+{
+	if (xen_pv_domain())
+		return virt_to_machine(&vmcoreinfo_note).maddr;
+	else
+		return __pa((unsigned long)(char *)&vmcoreinfo_note);
+}
+#endif /* CONFIG_KEXEC_CORE */
